### ISSUE TITLE
Error Building with CMake

### ISSUE THREAD

**Issue by simon-on-github** [kecsap/tensorflow_cpp_packaging#15]
**tl;dr:**
the library `png16` doesn't get found.
I tried to install libpng16-dev, but that resulted in a linking error

prerequisites:
1. Ubuntu 16.04
2. The package tensorflow-cpp-generic_1.13.1fix2.git20190225.6612da8951.ppa1_amd64.deb
3. Noted dependencies (libdouble-conversion-dev libfarmhash-dev libre2-dev libgif-dev libpng-dev libsqlite3-dev libsnappy-dev liblmdb-dev libicu-dev libjsoncpp-dev)
**How i installed the package:**

`sudo apt install ./tensorflow-cpp-generic_1.13.1fix2.git20190225.6612da8951.ppa1_amd64.deb`
This gives an dependency error (libpng-dev) , so I installed it (Apt notes that it installed libpng12-dev).
Then i installed the package again, but this time gives the error:
`N: Can't drop privileges for downloading as file '/home/simon/Downloads/tensorflow-cpp-generic_1.13.1fix2.git20190225.6612da8951.ppa1_amd64.deb' couldn't be accessed by user '_apt'. - pkgAcquire::Run (13: Permission denied)`
but the files get installed.

**Project setup**:

example.cpp:
```
#include <tensorflow/core/protobuf/meta_graph.pb.h>
#include <tensorflow/core/public/session.h>

int main() { }
```
CMakeLists.txt;
```
cmake_minimum_required(VERSION 3.5)

set (CMAKE_CXX_STANDARD 11)

set(CMAKE_PREFIX_PATH /usr/lib/cmake/tensorflow-cpp/ )

FIND_PACKAGE(TensorFlowCpp REQUIRED PATHS /usr/lib/cmake/tensorflow-cpp/)

INCLUDE_DIRECTORIES(${TENSORFLOWCPP_INCLUDE_DIRS})

ADD_EXECUTABLE(example example.cpp)

TARGET_LINK_LIBRARIES(example ${TENSORFLOWCPP_LIBRARIES})
```

**Error:**
Then building the dummy project This error appears:
```
[ 50%] Building CXX object CMakeFiles/example.dir/example.cpp.o
[100%] Linking CXX executable example
/usr/bin/ld: cannot find -lpng16
collect2: error: ld returned 1 exit status
CMakeFiles/example.dir/build.make:97: recipe for target 'example' failed
make[2]: *** [example] Error 1
CMakeFiles/Makefile2:67: recipe for target 'CMakeFiles/example.dir/all' failed
make[1]: *** [CMakeFiles/example.dir/all] Error 2
Makefile:83: recipe for target 'all' failed
make: *** [all] Error 2
```

**Informations:**
```
ldconfig -p | grep png

	libpng12.so.0 (libc6,x86-64) => /lib/x86_64-linux-gnu/libpng12.so.0
	libpng12.so.0 (libc6,x86-64) => /usr/lib/x86_64-linux-gnu/libpng12.so.0
	libpng12.so.0 (libc6,x86-64) => /usr/lib/libpng12.so.0
	libpng12.so (libc6,x86-64) => /usr/lib/x86_64-linux-gnu/libpng12.so
```
```
apt list --installed | grep png

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

libpng12-0/xenial-updates,xenial-security,now 1.2.54-1ubuntu1.1 amd64 [installed]
libpng12-dev/xenial-updates,xenial-security,now 1.2.54-1ubuntu1.1 amd64 [installed]
```

**Tried solutions:**
Since the png16 library can't be found and ubuntu uses libpng12-dev package for the libpng package, i tried to install the libpng16-dev package instead of the pnglib-dev package.
Then installing the libpng16-dev packge, the tensorflow and libpng12-dev package gets removed( :smile: )
So i installed the tensorflow package again, without the libpng-dev dependency.
`sudo dpkg -i --ignore-depends=libpng-dev tensorflow-cpp-generic_1.13.1fix2.git20190225.6612da8951.ppa1_amd64.deb`
Then building now, the following linking error appears.
**Error:**
```
simon:~/Downloads$ make
[ 50%] Linking CXX executable example
/usr/lib/tensorflow-cpp/libtensorflow-core.so: undefined reference to `ucnv_setToUCallBack_60'
/usr/lib/tensorflow-cpp/libtensorflow-core.so: undefined reference to `icu_60::UnicodeString::append(int)'
/usr/lib/tensorflow-cpp/libtensorflow-core.so: undefined reference to `icu_60::UnicodeStringAppendable::~UnicodeStringAppendable()'
/usr/lib/tensorflow-cpp/libtensorflow-core.so: undefined reference to `icu_60::UnicodeString::~UnicodeString()'
/usr/lib/tensorflow-cpp/libtensorflow-core.so: undefined reference to `icu_60::UnicodeString::toUTF8(icu_60::ByteSink&) const'
/usr/lib/tensorflow-cpp/libtensorflow-core.so: undefined reference to `std::thread::_State::~_State()@GLIBCXX_3.4.22'
/usr/lib/tensorflow-cpp/libtensorflow-core.so: undefined reference to `ucnv_close_60'
/usr/lib/tensorflow-cpp/libtensorflow-core.so: undefined reference to `ucnv_reset_60'
/usr/lib/tensorflow-cpp/libtensorflow-core.so: undefined reference to `std::thread::_M_start_thread(std::unique_ptr<std::thread::_State, std::default_delete<std::thread::_State> >, void (*)())@GLIBCXX_3.4.22'
/usr/lib/tensorflow-cpp/libtensorflow-core.so: undefined reference to `icu_60::StringCharacterIterator::~StringCharacterIterator()'
/usr/lib/tensorflow-cpp/libtensorflow-core.so: undefined reference to `icu_60::ErrorCode::~ErrorCode()'
/usr/lib/tensorflow-cpp/libtensorflow-core.so: undefined reference to `icu_60::UCharCharacterIterator::next32PostInc()'
/usr/lib/tensorflow-cpp/libtensorflow-core.so: undefined reference to `icu_60::UCharCharacterIterator::hasNext()'
/usr/lib/tensorflow-cpp/libtensorflow-core.so: undefined reference to `ucnv_getNextUChar_60'
/usr/lib/tensorflow-cpp/libtensorflow-core.so: undefined reference to `powf@GLIBC_2.27'
/usr/lib/tensorflow-cpp/libtensorflow-core.so: undefined reference to `ucnv_open_60'
/usr/lib/tensorflow-cpp/libtensorflow-core.so: undefined reference to `vtable for icu_60::UnicodeString'
/usr/lib/tensorflow-cpp/libtensorflow-core.so: undefined reference to `icu_60::ByteSink::GetAppendBuffer(int, int, char*, int, int*)'
/usr/lib/tensorflow-cpp/libtensorflow-core.so: undefined reference to `UCNV_TO_U_CALLBACK_SUBSTITUTE_60'
/usr/lib/tensorflow-cpp/libtensorflow-core.so: undefined reference to `typeinfo for icu_60::ByteSink'
/usr/lib/tensorflow-cpp/libtensorflow-core.so: undefined reference to `typeinfo for std::thread::_State@GLIBCXX_3.4.22'
/usr/lib/tensorflow-cpp/libtensorflow-core.so: undefined reference to `icu_60::ErrorCode::reset()'
/usr/lib/tensorflow-cpp/libtensorflow-core.so: undefined reference to `vtable for icu_60::ErrorCode'
/usr/lib/tensorflow-cpp/libtensorflow-core.so: undefined reference to `vtable for icu_60::UnicodeStringAppendable'
/usr/lib/tensorflow-cpp/libtensorflow-core.so: undefined reference to `icu_60::UMemory::operator delete(void*)'
/usr/lib/tensorflow-cpp/libtensorflow-core.so: undefined reference to `icu_60::ByteSink::Flush()'
/usr/lib/tensorflow-cpp/libtensorflow-core.so: undefined reference to `icu_60::UnicodeStringAppendable::appendCodePoint(int)'
/usr/lib/tensorflow-cpp/libtensorflow-core.so: undefined reference to `icu_60::UnicodeString::countChar32(int, int) const'
/usr/lib/tensorflow-cpp/libtensorflow-core.so: undefined reference to `icu_60::StringCharacterIterator::StringCharacterIterator(icu_60::UnicodeString const&)'
/usr/lib/tensorflow-cpp/libtensorflow-core.so: undefined reference to `expf@GLIBC_2.27'
/usr/lib/tensorflow-cpp/libtensorflow-core.so: undefined reference to `icu_60::ByteSink::~ByteSink()'
/usr/lib/tensorflow-cpp/libtensorflow-core.so: undefined reference to `uscript_getScript_60'
/usr/lib/tensorflow-cpp/libtensorflow-core.so: undefined reference to `logf@GLIBC_2.27'
collect2: error: ld returned 1 exit status
CMakeFiles/example.dir/build.make:97: recipe for target 'example' failed
make[2]: *** [example] Error 1
CMakeFiles/Makefile2:67: recipe for target 'CMakeFiles/example.dir/all' failed
make[1]: *** [CMakeFiles/example.dir/all] Error 2
Makefile:83: recipe for target 'all' failed
make: *** [all] Error 2
```
**Error informations:**:
```libpng12-0/xenial-updates,xenial-security,now 1.2.54-1ubuntu1.1 amd64 [installed]
libpng16-16/xenial,now 1.6.20-2 amd64 [installed,automatic]
libpng16-dev/xenial,now 1.6.20-2 amd64 [installed]
libpng16-devtools/xenial,now 1.6.20-2 amd64 [installed,automatic]
libpng16-tools/xenial,now 1.6.20-2 amd64 [installed,automatic]
```
```
ldconfig -p | grep png
	libpng16.so.16 (libc6,x86-64) => /lib/x86_64-linux-gnu/libpng16.so.16
	libpng16.so.16 (libc6,x86-64) => /usr/lib/x86_64-linux-gnu/libpng16.so.16
	libpng16.so.16 (libc6,x86-64) => /usr/lib/libpng16.so.16
	libpng16.so (libc6,x86-64) => /usr/lib/x86_64-linux-gnu/libpng16.so
	libpng12.so.0 (libc6,x86-64) => /lib/x86_64-linux-gnu/libpng12.so.0
	libpng12.so.0 (libc6,x86-64) => /usr/lib/x86_64-linux-gnu/libpng12.so.0
```

**Comment by simon-on-github** [comment]
i build tensorflow from source

**Comment by kecsap** [comment]
You can build Tensorflow from source, if you want to get a working package of TF 1.13.1, but the package you tried was built for Ubuntu Bionic while your system is Ubuntu Xenial. That is the explanation of the linker errors. The last version what I built for Xenial and published was TF 1.7:

https://github.com/kecsap/tensorflow_cpp_packaging/releases/tag/Tensorflow_1.7.0_git20180411_079539b

I don't release packages and maintain the build for earlier Ubuntu LTS versions because it is already complicated to keep it up with the latest Tensorflow version which evolves quickly. Currently, the problem with the latest TF 1.14+-2.0 versions that they broke/deprecate the usual model serving process (graph freezing -> load the frozen graph in C++). I don't know if it is possible to release any further version of Tensorflow. I can even build TF 2.0, but the graph loading dies in the protobuf serialization.

**Comment by simon-on-github** [comment]
Ok, thank you for the explanation
